net/socket.h

```diff
@@ -21,81 +21,217 @@ limitations under the License.
 LogBuffer& operator << (LogBuffer& log, const sockaddr_in& addr);
+LogBuffer& operator << (LogBuffer& log, const in6_addr& iaddr);
+LogBuffer& operator << (LogBuffer& log, const sockaddr_in6& addr);
 LogBuffer& operator << (LogBuffer& log, const sockaddr& addr);
+
 namespace photon {
-namespace net
-{
-    union IPAddr
-    {
-        uint32_t addr = 0;
-        struct { uint8_t a, b, c, d; };
-        explicit IPAddr(uint32_t nl)
-        {
-            from_nl(nl);
-        }
-        explicit IPAddr(const char* s)
-        {
-            struct in_addr addr;
-            if (inet_aton(s, &addr) == 0)
-                return;  // invalid IPv4 address
-            from_nl(addr.s_addr);
-        }
-        IPAddr() = default;
-        IPAddr(const IPAddr& rhs) = default;
-        uint32_t to_nl() const
-        {
-            return htonl(addr);
-        }
-        void from_nl(uint32_t nl)
-        {
-            addr = ntohl(nl);
+namespace net {
+
+    struct __attribute__ ((packed)) IPAddr {
+    public:
+        union {
+            in6_addr addr = {};
+            struct { uint16_t _1, _2, _3, _4, _5, _6; uint8_t a, b, c, d; };
+        };
+        // For compatibility, the default constructor is still 0.0.0.0 (IPv4)
+        IPAddr() {
+            map_v4(htonl(INADDR_ANY));
+        }
+        // V6 constructor (Internet Address)
+        explicit IPAddr(in6_addr internet_addr) {
+            addr = internet_addr;
+        }
+        // V6 constructor (Network byte order)
+        IPAddr(uint32_t nl1, uint32_t nl2, uint32_t nl3, uint32_t nl4) {
+            addr._in_addr_field[0] = nl1;
+            addr._in_addr_field[1] = nl2;
+            addr._in_addr_field[2] = nl3;
+            addr._in_addr_field[3] = nl4;
+        }
+        // V4 constructor (Internet Address)
+        explicit IPAddr(in_addr internet_addr) {
+            map_v4(internet_addr);
+        }
+        // V4 constructor (Network byte order)
+        explicit IPAddr(uint32_t nl) {
+            map_v4(nl);
+        }
+        // String constructor
+        explicit IPAddr(const char* s) {
+            if (inet_pton(AF_INET6, s, &addr) > 0) {
+                return;
+            }
+            in_addr v4_addr;
+            if (inet_pton(AF_INET, s, &v4_addr) > 0) {
+                map_v4(v4_addr);
+                return;
+            }
+            // Invalid string, make it a default value
+            *this = IPAddr();
+        }
+        // Check if it's actually an IPv4 address mapped in IPV6
+        bool is_ipv4() const {
+            if (ntohl(addr._in_addr_field[2]) != 0x0000ffff) {
+                return false;
+            }
+            if (addr._in_addr_field[0] != 0 || addr._in_addr_field[1] != 0) {
+                return false;
+            }
+            return true;
+        }
+        // We regard the default IPv4 0.0.0.0 as undefined
+        bool undefined() const {
+            return *this == V4Any();
+        }
```

Comment:
A default IPAddr is V4Any 0.0.0.0, which is also `undefined`. (used to be called empty)